### ISSUE TITLE
Charts broken in Android version 1.7

### ISSUE THREAD

**Issue by lolodomo** [openhab/openhab.android#119]
Charts are no more displayed in version 1.7 plus trying to display them seems to introduce terrible latency in the application as if the application was waiting for information it never obtain.

Here is a part of my sitemap that is working well  in the classic UI (on my PC) and with the old Android application running on my Android phone. But it is not working with the new Android application running on my Nexus 10 Android tablet. My tablet is running Android 5.1.1.

		Text label="Températures" {
			Text item=TemperatureSalon valuecolor=[<18="red",<19.5="orange"]
			Text item=TemperatureVeranda valuecolor=[<18="red",<19.5="orange"]
			Text item=TemperatureBureau valuecolor=[<18="red",<19.5="orange"]
			Text item=TemperatureChambre valuecolor=[<18="red",<19.5="orange"]
			Text item=TemperatureCinema valuecolor=[<18="red",<19.5="orange"]

			Switch item=RadiateurBureau mappings=[0="Confort", 1="Réduit", 2="Hors-gel", 3="Arrêt"]
			Switch item=RadiateurCinema mappings=[0="Confort", 1="Réduit", 2="Hors-gel", 3="Arrêt"]

			Switch item=TempChart1Periode mappings=[0="4 heures", 1="12 heures", 2="Jour", 3="Semaine"]
			Chart item=GTemperatureInt period=4h refresh=5000 visibility=[TempChart1Periode==0, TempChart1Periode=="Uninitialized"]
			Chart item=GTemperatureInt period=12h refresh=10000 visibility=[TempChart1Periode==1]
			Chart item=GTemperatureInt period=D refresh=30000 visibility=[TempChart1Periode==2]
			Chart item=GTemperatureInt period=W refresh=30000 visibility=[TempChart1Periode==3]

			Switch item=TempChart3Periode mappings=[0="Jour", 1="Semaine"]
			Chart item=GTemperatureVeranda period=D refresh=30000 visibility=[TempChart3Periode==0, TempChart3Periode=="Uninitialized"]
			Chart item=GTemperatureVeranda period=W refresh=30000 visibility=[TempChart3Periode==1]

			Text item=CO2Salon

			Text item=HumiditeSalon
			Text item=HumiditeVeranda
			Text item=HumiditeBureau
			Text item=HumiditeChambre
			Text item=HumiditeCinema

			Switch item=HumChart1Periode mappings=[0="Jour", 1="Semaine"]
			Chart item=GHumiditeInt period=D refresh=30000 visibility=[HumChart1Periode==0, HumChart1Periode=="Uninitialized"]
			Chart item=GHumiditeInt period=W refresh=30000 visibility=[HumChart1Periode==1]
		}

After commenting all stuff relative to charts in my sitemap, the Android application 1.7 is now working without any problem. So I assume there is a problem that is clearly relative to charts.

**Comment by belovictor** [comment]
Can you check it with a more simple page with just one static chart, just to be sure if charts are the broken thing?

**Comment by lolodomo** [comment]
I tried adding only that in my working sitemap:

		Text label="Test chart" {
			Chart item=GTemperatureInt period=12h refresh=10000
		}

and it does not work. When I click on "Test chart", the chart is not displayed even after waiting several seconds. And then when I click elsewhere to return to another page, it takes a long time to display data (while it is faster/normal when I don't come form the chart page.

**Comment by belovictor** [comment]
Strange, everything is fine in my installation. Which version of openHAB on which platform do you use, which exact Android app version? Is there a difference in Android version on your phone and Nexus?

**Comment by lolodomo** [comment]
openHAB 1.7 on RPI
Android 4.4.2 + HABdroid 1.5.1.1 = OK
Android 5.1.1 + HABdroid 1.7.0.1 = KO

**Comment by belovictor** [comment]
Can you get logcat from your android 5.1.1 device when entering the page with charts?

**Comment by elisman** [comment]
@lolodomo : Do you have the security option on, off or external? And do the charts work in your local network?

**Comment by lolodomo** [comment]
Yes, my security option is set to ON.
The charts do not work even in my local network.

**Comment by elisman** [comment]
@lolodomo : Try to set security option to OFF (only for testing purposes) or to EXTERNAL for tests within your local network. My Charts work in my local network with the security option set to EXTERNAL. Keep in mind that you need to restart the openHAB service after changing the security option in order for this to take effect.

Please let us know if this works.

**Comment by lolodomo** [comment]
I set security option to OFF and it changed nothing.
Note that I did my test after upgrading server to 1.7.1.
BUT I have additional information with errors in server:logs:


2015-08-01 06:23:21.832 [WARN ] [e.jetty.servlet.ServletHandler] - Error for /chart
java.lang.OutOfMemoryError: Java heap space
        at java.awt.image.DataBufferInt.<init>(DataBufferInt.java:75) ~[na:1.8.0]
        at java.awt.image.Raster.createPackedRaster(Raster.java:467) ~[na:1.8.0]
        at java.awt.image.DirectColorModel.createCompatibleWritableRaster(DirectColorModel.java:1032) ~[na:1.8.0]
        at java.awt.image.BufferedImage.<init>(BufferedImage.java:340) ~[na:1.8.0]
        at org.openhab.ui.internal.chart.DefaultChartProvider.createChart(DefaultChartProvider.java:233) ~[na:na]
        at org.openhab.ui.internal.chart.ChartServlet.doGet(ChartServlet.java:245) ~[na:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:735) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:848) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61) ~[na:na]
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128) ~[na:na]
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60) ~[na:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:848) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:598) ~[org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:486) ~[org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231) ~[na:na]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1065) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:413) ~[org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:192) ~[na:na]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:999) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:250) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:149) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:111) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.Server.handle(Server.java:350) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:454) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:890) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:944) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:630) ~[na:na]
        at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:230) ~[na:na]
        at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:77) ~[org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:606) ~[na:na]
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:46) ~[na:na]

Other error:


2015-08-01 06:26:09.650 [WARN ] [e.jetty.servlet.ServletHandler] - /chart
java.lang.IndexOutOfBoundsException: null
        at javax.imageio.stream.FileCacheImageOutputStream.seek(FileCacheImageOutputStream.java:170) ~[na:1.8.0]
        at javax.imageio.stream.FileCacheImageOutputStream.close(FileCacheImageOutputStream.java:231) ~[na:1.8.0]
        at javax.imageio.ImageIO.write(ImageIO.java:1580) ~[na:1.8.0]
        at org.openhab.ui.internal.chart.ChartServlet.doGet(ChartServlet.java:247) ~[na:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:735) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:848) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61) ~[na:na]
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128) ~[na:na]
        at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60) ~[na:na]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:848) ~[javax.servlet_3.0.0.v201112011016.jar:na]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:598) ~[org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:486) [org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1065) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:413) [org.eclipse.jetty.servlet_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:192) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:999) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:250) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:149) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:111) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.Server.handle(Server.java:350) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:454) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:890) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:944) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:630) [org.eclipse.jetty.http_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:230) [org.eclipse.jetty.http_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:77) [org.eclipse.jetty.server_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:606) [org.eclipse.jetty.io_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:46) [org.eclipse.jetty.io_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:603) [org.eclipse.jetty.util_8.1.3.v20120522.jar:8.1.3.v20120522]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:538) [org.eclipse.jetty.util_8.1.3.v20120522.jar:8.1.3.v20120522]
        at java.lang.Thread.run(Thread.java:744) [na:1.8.0]

My server is running on a RPI.

Hope it could help.

**Comment by lolodomo** [comment]
I tried again after few hours my server is running with security set to ON.
In this case, I have no error in the server logs but no chart displayed in HABdroid 1.7 (no problem with old version of HABdroid and classic ui).

I will try again tomorrow with security set to OFF to be sure that the logs I provided today were not due only to my database that was empty (I cleared everything when I installed server 1.7.1). By the way, these errors are not normal.

**Comment by lolodomo** [comment]
Ok, I was able again to reproduce the java.lang.IndexOutOfBoundsException: null with HABdroid 1.7. But the scene to get it is not very clear for me...

Then I tried again with security set to ON and this time it gets more time to generate errors but I finally get in the server a serial of 

[WARN ] [e.jetty.servlet.ServletHandler] - Error for /chart
java.lang.OutOfMemoryError: Java heap space

Apparently it is at this time HABdroid becomes very slow, probably waiting for the server ?

During this test session, I saw one time a chart displayed. I am not sure it was the expected display, the legend was very very small and lines were very thin too.
